### ISSUE TITLE
Add a signature lookup

### ISSUE THREAD

**Issue by tsuza** [Hpmason/retour-utils#13]
Right now we can retrieve a function via symbol and offset. It'd be useful to be able to retrieve it also via a signature, which makes detouring much safer and less prone to breakage in programs that still get updated.

It could support both IDA-style signatures ( e.g. `55 8B EC 83 EC 0C 56 57 8B F9 E8 ?? ?? ?? ?? 8B F0 85 F6 0F 84 ?? ?? ?? ?? 8B 16` ) and normal-style signatures ( I have no clue what they're supposed to be called ) ( e.g. `\x55\x8B\xEC\x83\xEC\x0C\x56\x57\x8B\xF9\xE8\x2A\x2A\x2A\x2A\x8B\xF0\x85\xF6\x0F\x84\x2A\x2A\x2A\x2A\x8B\x16` ).

It would look something along the lines of:
```rust
#[hook_module("lua52.dll")]
mod lua {
    // #[hook_module] will create this
    // const MODULE_NAME: &str = "lua52.dll"
    // and
    // pub unsafe init_detours() -> crate::Result<()> {..}
    // which will initialize all the StaticDetours generated by the macro inside this module

    #[allow(non_camel_case_types)]
    type lua_State = ();
    #[allow(non_camel_case_types)]
    type lua_Alloc = ();
    
    // Creates a StaticDetour called Lua_newstate with the same function type as our function 
    // (minus abi/unsafe to work with retour crate)
    #[hook(unsafe extern "C" Lua_newstate, signature = "\x55\x8B\xEC\x83\xEC\x0C\x56\x57\x8B\xF9\xE8\x2A\x2A\x2A\x2A\x8B\xF0\x85\xF6\x0F\x84\x2A\x2A\x2A\x2A\x8B\x16")]
    pub fn newstate(f: *mut lua_Alloc, ud: *mut std::ffi::c_void) -> *mut lua_State {
        unsafe {
            Lua_newstate.call(f, ud)
        }
    }
    // More lua hooks
}
```

I would try to implement this if it weren't for the fact that I have no experience with macros. The most I could do is implement a signature lookup.

**Comment by Hpmason** [comment]
I wasn't aware of signature lookup method, though something like that would be extremely helpful. 

Getting the signature from the macro wouldn't be difficult, so if you'd be able to write the signature lookup, I could help you get the signature from the macro.

**Comment by tsuza** [comment]
Yeah, it's pretty useful over raw offsets. I'll look into implementing it this weekend. Thanks!

**Comment by tsuza** [comment]
I haven't forgotten about this! I have just gotten busy with classes and now exams. Unfortunately I won't be able to work on it until the first few days of March. Sorry!
